server/events/workflow_hooks_command_runner.go

```diff
@@ -0,0 +1,144 @@
+}
+
+// RunPreHooks runs pre_workflow_hooks when PR is opened or updated.
+func (w *DefaultWorkflowHooksCommandRunner) RunPreHooks(
+	baseRepo models.Repo,
+	headRepo models.Repo,
+	pull models.PullRequest,
+	user models.User,
+) {
+	if opStarted := w.Drainer.StartOp(); !opStarted {
+		if commentErr := w.VCSClient.CreateComment(baseRepo, pull.Num, ShutdownComment, "pre_workflow_hooks"); commentErr != nil {
+			w.Logger.Log(logging.Error, "unable to comment that Atlantis is shutting down: %s", commentErr)
+		}
+		return
+	}
+	defer w.Drainer.OpDone()
+
+	log := w.buildLogger(baseRepo.FullName, pull.Num)
+	defer w.logPanics(baseRepo, pull.Num, log)
+
+	log.Info("Running Pre Hooks for repo: ")
+
+	unlockFn, err := w.WorkingDirLocker.TryLock(baseRepo.FullName, pull.Num, DefaultWorkspace)
+	if err != nil {
+		log.Warn("workspace is locked")
+		return
+	}
+	log.Debug("got workspace lock")
+	defer unlockFn()
+
+	repoDir, _, err := w.WorkingDir.Clone(log, headRepo, pull, DefaultWorkspace)
+	if err != nil {
+		log.Err("unable to run pre workflow hooks: %s", err)
+		return
+	}
+
+	workflowHooks := make([]*valid.WorkflowHook, 0)
+	for _, repo := range w.GlobalCfg.Repos {
+		if repo.IDMatches(baseRepo.ID()) && len(repo.WorkflowHooks) > 0 {
+			workflowHooks = append(workflowHooks, repo.WorkflowHooks...)
+		}
+	}
+
+	ctx := models.WorkflowHookCommandContext{
+		BaseRepo: baseRepo,
+		HeadRepo: headRepo,
+		Log:      log,
+		Pull:     pull,
+		User:     user,
+		Verbose:  false,
+	}
+
+	result := w.runHooks(ctx, workflowHooks, repoDir)
+
+	if result.HasErrors() {
+		log.Err("pre workflow hook run error results: %s", result.Errors())
+	}
+}
+
+func (w *DefaultWorkflowHooksCommandRunner) runHooks(
+	ctx models.WorkflowHookCommandContext,
+	workflowHooks []*valid.WorkflowHook,
+	repoDir string,
+) *WorkflowHooksCommandResult {
+	result := &WorkflowHooksCommandResult{
+		WorkflowHookResults: make([]models.WorkflowHookResult, 0),
+	}
+	for _, hook := range workflowHooks {
+		out, err := w.WorkflowHookRunner.Run(ctx, hook.RunCommand, repoDir)
+
+		res := models.WorkflowHookResult{
+			Output: out,
+		}
+
+		if err != nil {
+			res.Error = err
+			res.Success = false
+		} else {
+			res.Success = true
+		}
+
+		result.WorkflowHookResults = append(result.WorkflowHookResults, res)
+
+		if !res.IsSuccessful() {
+			return result
+		}
+	}
+
+	return result
+}
+
+func (w *DefaultWorkflowHooksCommandRunner) buildLogger(repoFullName string, pullNum int) *logging.SimpleLogger {
+	src := fmt.Sprintf("%s#%d", repoFullName, pullNum)
+	return w.Logger.NewLogger(src, true, w.Logger.GetLevel())
+}
+
+// logPanics logs and creates a comment on the pull request for panics.
+func (w *DefaultWorkflowHooksCommandRunner) logPanics(baseRepo models.Repo, pullNum int, logger logging.SimpleLogging) {
+	if err := recover(); err != nil {
+		stack := recovery.Stack(3)
+		logger.Err("PANIC: %s\n%s", err, stack)
+		if commentErr := w.VCSClient.CreateComment(
```

Comment:
Seems like a bad user experience to do that but don't want to deviate from the standard if that's set already.

Reply:
Yep, it is in the command_runner.go both `RunCommentCommand` and `RunAutoplanCommand` use it. What would alternative be? Potentially we can log and move on for pre_workflow_hooks.

Reply:
It's a good question.

i think in this case, the hooks are going to be only configured server side right? So customers shouldn't really know about it. Hence I would lean towards something like we just log an error and a stat (and alarm on this internally) and just move on.

In our case, there will probably be some downstream error about missing an atlantis configuration file so it's probably ok. In the general case, I'm not sure what people would use this for, but might be safer to not panic on failure in case it's something non-critical.

What do you think?

Reply:
I think that should work for our use case.